Mathlib/CategoryTheory/Pi/Basic.lean

```diff
@@ -0,0 +1,326 @@
+and the identity functor. -/
+@[simps]
+def comapId : comap C (id : I → I) ≅ 𝟭 (∀ i, C i)
+    where
+  hom := { app := fun X => 𝟙 X, naturality := by simp only [comap]; aesop_cat}
+  inv := { app := fun X => 𝟙 X }
+#align category_theory.pi.comap_id CategoryTheory.Pi.comapId
+
+example (g : J → I) : (j : J) → Category (C (g j)) := by infer_instance 
+
+variable {I}
+
+variable {K : Type w₂}
+
+/-- The natural isomorphism comparing between
+pulling back along two successive functions, and
+pulling back along their composition
+-/
+@[simps!]
+def comapComp (f : K → J) (g : J → I) : comap C g ⋙ comap (C ∘ g) f ≅ comap C (g ∘ f)
+    where
+  hom := { 
+    app := fun X b => 𝟙 (X (g (f b))) 
+    naturality := fun X Y f' => by simp only [comap,Function.comp]; funext; simp
+    }
+  inv := { 
+    app := fun X b => 𝟙 (X (g (f b)))
+    naturality := fun X Y f' => by simp only [comap,Function.comp]; funext; simp
+    }
+  hom_inv_id := by 
+    simp only [comap] 
+    ext Y 
+    simp [CategoryStruct.comp,CategoryStruct.id]
+  inv_hom_id := by 
+    simp only [comap]
+    ext X
+    simp [CategoryStruct.comp,CategoryStruct.id]
+#align category_theory.pi.comap_comp CategoryTheory.Pi.comapComp
+
+/-- The natural isomorphism between pulling back then evaluating, and just evaluating. -/
+@[simps!]
+def comapEvalIsoEval (h : J → I) (j : J) : comap C h ⋙ eval (C ∘ h) j ≅ eval C (h j) :=
+  NatIso.ofComponents (fun f => Iso.refl _) (by simp only [Iso.refl]; aesop_cat)
+#align category_theory.pi.comap_eval_iso_eval CategoryTheory.Pi.comapEvalIsoEval
+
+end
+
+section
+
+variable {J : Type w₀} {D : J → Type u₁} [∀ j, Category.{v₁} (D j)]
+
+/- Porting note: We have the following warning but it just looks 
+to me to be permutation of the universe labels 
+
+warning: category_theory.pi.sum_elim_category ->
+\  CategoryTheory.Pi.sumElimCategory is a dubious translation:
+lean 3 declaration is forall {I : Type.{u1}} (C : I -> Type.{u3}) [_inst_1 :
+forall (i : I), CategoryTheory.Category.{u2, u3} (C i)] {J : Type.{u1}} {D : J
+-> Type.{u3}} [_inst_2 : forall (j : J), CategoryTheory.Category.{u2, u3} (D
+j)] (s : Sum.{u1, u1} I J), CategoryTheory.Category.{u2, u3} (Sum.elim.{u1, u1,
+succ (succ u3)} I J Type.{u3} C D s) but is expected to have type forall {I :
+Type.{u3}} (C : I -> Type.{u1}) [_inst_1 : forall (i : I),
+CategoryTheory.Category.{u2, u1} (C i)] {J : Type.{u3}} {D : J -> Type.{u1}}
+[_inst_2 : forall (j : J), CategoryTheory.Category.{u2, u1} (D j)] (s :
+Sum.{u3, u3} I J), CategoryTheory.Category.{u2, u1} (Sum.elim.{u3, u3, succ
+(succ u1)} I J Type.{u1} C D s) Case conversion may be inaccurate. Consider
+using '#align category_theory.pi.sum_elim_category
+CategoryTheory.Pi.sumElimCategoryₓ'. -/
+instance sumElimCategory : ∀ s : Sum I J, Category.{v₁} (Sum.elim C D s)
+  | Sum.inl i => by
+    dsimp
+    infer_instance
+  | Sum.inr j => by
+    dsimp
+    infer_instance
+#align category_theory.pi.sum_elim_category CategoryTheory.Pi.sumElimCategory
```

Comment:
```suggestion
instance sumElimCategory : ∀ s : Sum I J, Category.{v₁} (Sum.elim C D s)
  | Sum.inl i => by
    dsimp
    infer_instance
  | Sum.inr j => by
    dsimp
    infer_instance
#align category_theory.pi.sum_elim_category CategoryTheory.Pi.sumElimCategoryₓ -- Porting note: reordered universes
```

Reply:
@jcommelin they look the same to me up to relabeling. I just wanted another set of eyes to confirm. 😆 

If you agree, I will delete all of this.

Reply:
I agree. But note that my suggestion adds the little `ₓ` to the align statement, as the comment instructs. And please leave a porting note.

Reply:
Done.